File: Problemset/find-majority-element-lcci/find-majority-element-lcci.py

# @Title: 主要元素 (Find Majority Element LCCI)
# @Author: cocofe
# @Date: 2020-10-28 22:58:28
# @Runtime: 16 ms
# @Memory: 14.1 MB

class Solution(object):
    def majorityElement(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        length = len(nums)
        nums = sorted(nums)
        mid = nums[length / 2]
        mid_count = 0
        for num in nums:
            if num == mid:
                mid_count += 1
        return mid if mid_count > length / 2 else -1

File: Problemset/find-majority-element-lcci/find-majority-element-lcci.java

// @Title: 主要元素 (Find Majority Element LCCI)
// @Author: cocofe
// @Date: 2021-07-10 03:10:42
// @Runtime: 1 ms
// @Memory: 44.5 MB

class Solution {
    public int majorityElement(int[] nums) {
        int aliveCount = 0;
        int alive = -1;
        for (int num : nums) {
            if (aliveCount == 0) {
                alive = num;
                aliveCount ++;
                continue;
            }
            if (alive == num) {
                aliveCount++;
            } else {
                aliveCount --;
            }
        }
        if (aliveCount == 0) {
            return -1;
        }
        int count = 0;
        for (int num : nums) {
            if (num == alive) {
                count ++;
            }
        }
        return count > nums.length / 2 ? alive : -1;

    }
}
